### PR TITLE
Add a "pub workspace" to the root of the engine repository

### PR DESCRIPTION
**Pull request by matanlurey**
... and use it in `engine_tool` to prove everything is working, i.e. on CI and elsewhere.

Partial work towards https://github.com/flutter/flutter/issues/147883.
Supersedes and closes https://github.com/flutter/engine/pull/51782 (which was a prototype).

See also:
- https://flutter.dev/go/pub-workspace, the design doc on the feature.
- https://github.com/dart-lang/pub-dev/pull/7762, an example PR.

I'll probably end up moving the inline docs in `pubspec.yaml` to a `docs/*.md` once that's a thing.

### CODE REVIEW

**Review by johnmccutchan — APPROVED** [review]

File: pubspec.yaml
Comment (jtmcdole) [inline pubspec.yaml:52]: s/manuallu/manually

File: pubspec.yaml
Comment (jtmcdole) [inline pubspec.yaml:146]: Why are there extra packages in here that aren't referenced in `dependencies` above?  e.g. `async` is listed here but not above.

File: tools/engine_tool/pubspec.yaml
Comment (jtmcdole) [inline tools/engine_tool/pubspec.yaml:31]: this `path` doesn't get picked up by the workspace?

**Review by jtmcdole — APPROVED** [review]

File: pubspec.yaml
Comment (matanlurey) [reply pubspec.yaml:146]: All _transitive_ dependencies have to also be listed. I can update the comments to make that clear.

(Think of this more like `//third_party/.../BUILD` than a traditional pubspec)

File: tools/engine_tool/pubspec.yaml
Comment (matanlurey) [reply tools/engine_tool/pubspec.yaml:31]: No, because `process_fakes` isn't a pub dependency (it's not on pub.dev) so there is no resolution to point to.

Comment (jtmcdole) [reply tools/engine_tool/pubspec.yaml:31]: filed: https://github.com/dart-lang/pub/issues/4318

File: pubspec.yaml
Comment (sigurdm) [inline pubspec.yaml:105]: You could enable: https://dart.dev/tools/linter-rules/sort_pub_dependencies

**Comment by chinmaygarde** [comment]
The issue this was blocked on has been resolved and rolled in. Does this just need a rebase?

**Comment by jtmcdole** [comment]
> The issue this was blocked on has been resolved and rolled in. Does this just need a rebase?

I think we're still looking at some options for overrides. See dart-lang/pub/issues/4318

File: pubspec.yaml
Comment (jtmcdole) [inline pubspec.yaml:115]: move `litetest` and `process_fakes` to `dependency_overrides` and the sub-packages don't need to include redundant paths.

**Comment by matanlurey** [comment]
I am still blocked on a new enough Dart SDK that supports `workspaces`.

Revving here: https://dart-review.googlesource.com/c/sdk/+/373903

File: pubspec.yaml
Comment (navaronbracke) [inline pubspec.yaml:23]: `by checking the are available` seems like a typo?

**Comment by chinmaygarde** [comment]
The linked pull request was merged. Are you still blocked on this?

File: pubspec.yaml
Comment (matanlurey) [reply pubspec.yaml]: Done.

File: pubspec.yaml
Comment (matanlurey) [reply pubspec.yaml:99]: Added a TODO for now, thanks!

File: pubspec.yaml
Comment (matanlurey) [reply pubspec.yaml]: Great suggestion, done!

File: pubspec.yaml
Comment (matanlurey) [reply pubspec.yaml]: Done!

File: tools/pub_get_offline.py
Comment (chinmaygarde) [inline tools/pub_get_offline.py:69]: I see [pyyaml in the DEPS](https://github.com/flutter/engine/blob/08058e331a43477a06e4a5613566c41765712803/DEPS#L669). Is that not usable?

Comment (matanlurey) [reply tools/pub_get_offline.py:82]: Sounds good, I used the same pattern as `gen_dart_package_config.py` and went with:

```py
#!/usr/bin/env vpython3

# [VPYTHON:BEGIN]
# python_version: "3.8"
# wheel <
#   name: "infra/python/wheels/pyyaml/${platform}_${py_python}_${py_abi}"
#   version: "version:5.4.1.chromium.1"
# >
# [VPYTHON:END]
```

Comment (zanderso) [reply tools/pub_get_offline.py:82]: We should only be fetching dependencies from the network from the DEPS file.

Comment (matanlurey) [reply tools/pub_get_offline.py:82]: Right, pubspec files with `resolution: workspace` do not fetch dependencies at all.

This script will throw saying "no package_config.json", hence this change.

Comment (zanderso) [reply tools/pub_get_offline.py:82]: Ah, sorry. My comment was in reference to the discussion of pyyaml and using the wheel.

File: pubspec.yaml
Comment (loic-sharma) [inline pubspec.yaml:23]:
```suggestion
# Dependencies need to be added either via the DEPS file, or by checking if they
```

File: pubspec.yaml
Comment (loic-sharma) [inline pubspec.yaml:27]: Is (4) the step you need to do? Consider something like:

```suggestion
# If dependencies are available, see (4) for how to add dependencies to a package in the workspace.
```

File: pubspec.yaml
Comment (loic-sharma) [inline pubspec.yaml:33]:
```suggestion
# header in its respective `pubspec.yaml`:
```

File: pubspec.yaml
Comment (loic-sharma) [inline pubspec.yaml:75]: These comments are phenomenal, excellent work!

File: DEPS
Comment (matanlurey) [reply DEPS:80]: Hmm, I think this was a bad merge, I don't recall intentionally wanting to change DEPS in this PR.

File: tools/pub_get_offline.py
Comment (matanlurey) [reply tools/pub_get_offline.py:157]: Hmm... I'm actually not sure. I _assume_ no, but let me give it a shot.

Comment (matanlurey) [reply tools/pub_get_offline.py:157]: Short answer: Yes, it's possible. It ends up adding it to the _root_ `.dart_tool/package_config.json`.

Long answer: Our script to protect against pub dependencies _still_ works, as it catches the root package config:

```diff
diff --git a/tools/engine_tool/lib/main.dart b/tools/engine_tool/lib/main.dart
index e524e07ad0..d2864ee1db 100644
--- a/tools/engine_tool/lib/main.dart
+++ b/tools/engine_tool/lib/main.dart
@@ -5,6 +5,7 @@
 import 'dart:ffi' as ffi show Abi;
 import 'dart:io' as io show Directory, Platform, exitCode, stderr;
 
+import 'package:archive/archive.dart';
 import 'package:engine_build_configs/engine_build_configs.dart';
 import 'package:engine_repo_tools/engine_repo_tools.dart';
 import 'package:path/path.dart' as p;
@@ -17,6 +18,7 @@ import 'src/logger.dart';
 import 'src/phone_home.dart';
 
 void main(List<String> args) async {
+  print(getAdler32);
   if (phoneHome(args)) {
     return;
   }
diff --git a/tools/engine_tool/pubspec.yaml b/tools/engine_tool/pubspec.yaml
index 1b71b9b688..83759735d5 100644
--- a/tools/engine_tool/pubspec.yaml
+++ b/tools/engine_tool/pubspec.yaml
@@ -13,6 +13,7 @@ environment:
 resolution: workspace
 
 dependencies:
+  archive: any
   args: any
   engine_build_configs: any
   engine_repo_tools: any
```

... and then:

```sh
matanl@matanl-macbookpro4 engine_tool % gclient sync
Syncing projects: 100% (143/143), done.
Running hooks:  46% ( 7/15) pub get --offline   
________ running 'python3 src/flutter/tools/pub_get_offline.py' in '/Users/matanl/Developer/engine'
Error: package "archive" was fetched from pub
Error: package "crypto" was fetched from pub
Error: package "typed_data" was fetched from pub
Error: 3 packages were fetched from pub for /Users/matanl/Developer/engine/src/flutter
Please fix the pubspec.yaml for /Users/matanl/Developer/engine/src/flutter so that all dependencies are path dependencies
Error: Command 'python3 src/flutter/tools/pub_get_offline.py' returned non-zero exit status 1 in /Users/matanl/Developer/engine
Error: package "archive" was fetched from pub
Error: package "crypto" was fetched from pub
Error: package "typed_data" was fetched from pub
Error: 3 packages were fetched from pub for /Users/matanl/Developer/engine/src/flutter
Please fix the pubspec.yaml for /Users/matanl/Developer/engine/src/flutter so that all dependencies are path dependencies
```

I do think the UX of allowing package-specific dependencies and silently adding them to the root pubspec is confusing, but not a blocker for this effort. I'll file an issue on dart-lang/pub and see what the team thinks in terms of UX.

Comment (matanlurey) [reply tools/pub_get_offline.py:157]: Done in https://github.com/dart-lang/pub/issues/4328.

Comment (zanderso) [reply tools/pub_get_offline.py:157]: Thanks for checking!

File: DEPS
Comment (zanderso) [inline DEPS:17]: More git merge troubles?

**Review by zanderso — APPROVED** [review]
lgtm modulo more git merge troubles?

**Review by jtmcdole — APPROVED** [review]
still lgtm

**Pull request merged**
